### DISCUSSION TITLE
QPD for focus lock system in ImSwitch

### DISCUSSION THREAD

**Discussion by jacopoabramo** [ImSwitch/ImSwitch#101] [Q&A] [answered]
From what I've read in the documentation, ImSwitch uses a camera as the input for the control loop which locks the Z-axis position, so I assume that at on the Managers layer the focus lock uses a DetectorManager for the control loop.

Now here's the situation: we were hoping to use instead a Quad Photodiode (QPD) which takes a laser input, redirected from the sample we're currently analyzing. So instead of having an image, we would have as an input for the control loop a set of X-Y coordinates from the QPD (which are calculated using the currents read from the QPD), thus a relation of (X, Y) -> Z. So basically just a number.

Can we integrate this into ImSwitch? Because our idea is to use an Arduino to read the coordinates and send them via USB to ImSwitch. But I'm unsure wether this is feasible to integrate in the widget - and if so, how.

If you need more details or my question is unclear let me know.

**Comment by xcasas** [comment, marked as answer]
Hi @jacopoabramo, 

I think you could do that by modifying the [FocusLockController](https://github.com/kasasxav/ImSwitch/blob/cb65c784c641f920f3060ed691c9b6292b135f1f/imswitch/imcontrol/controller/controllers/FocusLockController.py), and modify its entry in [SetupInfo](https://github.com/kasasxav/ImSwitch/blob/cb65c784c641f920f3060ed691c9b6292b135f1f/imswitch/imcontrol/model/SetupInfo.py). Maybe instead of having camera we can have something more general, like feedbackDevice, depends on how much it changes depending on the device (for a camera we do display the image, while maybe for the QPD you only want a 1D plot of the current). What do you think?

**Reply by jacopoabramo** [reply]
Understood. From a first glance it seems that instead of a ```grabCameraFrame()``` method there should be a more generic ```grabData()``` method which performs a read from the generic device. In my case it's basically an Arduino which computes the X-Y position of the laser on the QPD and translates this with a Z coordinate on the lock. I still have to figure out the details, but to be honest I entered holidays mode and I'm running around the office with a Santa hat y_y I understand your suggestion though. I'll try to follow up on that!

**Reply by xcasas** [reply]
Haha I love this !!! I think the grabData idea is excellent also! 
Should go get a santa hat for myself 🧐
